### DISCUSSION TITLE
Chunking seems not to be working when importing from packages in monorepo

### DISCUSSION THREAD

**Discussion by Poolshark** [webpack/webpack#15854] [Q&A] [answered]
I'm trying to figure this out since days now but I simply can't get my head around what's going on. I further don't know if this problem is **babel** related or if it originates from **webpack**.

I have a monorepo with **3 apps** which all rely on components from `packages`
```
frontend/
  apps/       <-- each app is a package in the scope of the Monroe
    app1/
    ...
  packages/   <-- this is a package in the monorepo scope
    components/
    providers/
    ...
```

Inside _app specific_ components I import eg. components like
```tsx
import { MyComponent } from 'packages/components';
```
Inside _packages_ I have aliased imports, eg. requesting a _component_ in _providers_
```tsx
import { MyComponent } from 'frontend/components';
```

The `tsconfig.json` in **packages** defines the local aliases
```
{
  ...
  "paths": {
    "frontend/components": ["./components"],
    ...
}
```

In my `webpack.config.ts` I define the aliases like
```ts
{
  resolve: {
    alias: {
      components: resolve(__dirname, "../../../frontend/apps/campus/src/components") // <-- resembles the absolute path to the config file
    }
  }
}
```

It all works but when I check my bundle I get a **main chunk** with the entire content from **packages**? Does that have something to do with my aliases or should I do something with my chunk settings? 

Here's a screenshot from the bundle analyzer:
![Screenshot 2022-05-20 at 13 18 16](https://user-images.githubusercontent.com/9102922/169517177-c01daa6b-e5f1-476a-b9ff-29ea0e183ed3.png)

Especially interesting is the `iconpack_default`, from which I **dynamically import** icons. I would have expected this to put every icon inside a separate chunk...

I know, this might be quite vague but I'd appreciate any kind of push in the right direction!

Thanks! 🙌

**Comment by vankop** [comment, marked as answer]
it looks like your imports example is not related to including all `packages`, you can try to use https://statoscope.tech/ to check what modules request `packages/*`

**Reply by Poolshark** [reply]
Hi @vankop!

Thanks for the answer! 🙌

Using https://statoscope.tech/ definitely brought some light into the dark. Since then I've done some intensive tests to try finding out what's going on. So far I found that:
- the issue does not occur from **path aliases**
- it has nothing to do with the **monorepo structure**

**BUT** it seems that webpack has somehow problems with `.tsx` files! As soon as I export more than one TSX component from an `index.ts` **ALL** the sources within that index get bundled! I'm pretty sure this is some kind of misconfig, right?

Just to be clear here - assume the following  `App.tsx` _(which is an import in the entry file `index.ts`)_
```ts
// App.tsx - types skipped for better readability
import { Test } from 'packages/test';

export const App = () => {
  return <Test />;
}
```

The "package" test looks like:
```
test/
  Test.tsx
  Test1.tsx
  testFunc.ts
  index.ts
```
where `index.ts` is
```ts
// index.ts
export { Test } from './Test';
export { Test1 } from './Test1';
export { testFunc } from './testFunc'
```

The **app bundle** with the above example includes `Test` AND `Test1`! I would expect it to only include `Test` since `Test1` is not in use. However, the more `.tsx` files I add and export from the `index.ts`, the more components get added to the bundle.

If I have only `.ts` files, I have no issue! Then only the source I am using inside `App.tsx` gets added to the bundle.

This might be a very stupid question, so again any push into the right direction welcome! 🙌

**Reply by vankop** [reply]
you should use [sideEffects flag](https://webpack.js.org/guides/tree-shaking/#clarifying-tree-shaking-and-sideeffects) to make this work, also make sure that this code doesn't transpiles to commonjs

**Reply by Poolshark** [reply]
Thank you!

Tested this but unfortunately it had no effect. Moreover, it seems that now **all `.ts` files which do include any React specific code** _(eg. hooks)_, will be included in the bundle. So

```ts
const test = () => {
  return "Hello World!";
}
```
is **not** included (unused), whereas
```ts
import { useEffect } from 'react';
const test = {
  useEffect(() => { 
    //do nothing 
  }, []);
  return "Hello World!";
}
```
is automatically included although not used in `App.tsx` 🙈

I'm sorry for this mess and I'm pretty sure I'm just missing a minor config here. Funny enough, if I change **all** my imports in `App.tsx` to the exact file position _(eg. `import { test } from 'packages/test'`)_ it works as expected. But the import via the `index.ts` causes all files to be included in the bundle.

**Reply by vankop** [reply]
> But the import via the index.ts causes all files to be included in the bundle.

As I said you need to set sideEffects flag correctly **or** result of ts transpiling is commonjs this prevents tree-shaking. You can try to create small reproducible repo I will take a look

**Reply by Poolshark** [reply]
Hey!

I have played around with the `sideEffects` flag but so far nothing did the job. Guess just setting it to `false` does not do the trick.

Anyway, I have managed to create a [minimal example](https://github.com/Poolshark/webpack-minimal-example) from my (quite complex) monorepo, which actually shows the problem. I hope this works for you!

@vankop I really appreciate the time you spending helping me out here! Thank you so much! 🙌

PS.: Let me know if the minimal example is sufficient.

**Reply by vankop** [reply]
you should set `optimization.sideEffects=true` https://github.com/Poolshark/webpack-minimal-example/blob/master/config/webpack/config/apps.webpack.ts#L107 this will enable side effects plugin on webpack side.
_____

This should help 
<img width="1069" alt="Screenshot 2022-05-24 at 13 38 25" src="https://user-images.githubusercontent.com/10380560/170015835-25aca1c8-ee40-47b9-b2e0-c8f8beb40a11.png">

**Reply by Poolshark** [reply]
Hi!

Thank you very much! This solved the problem! 🙌

It's just very weird that I had to set `sideEffetcs: false` in the `package.json` of the frontend package and `sideEffects: true` in `optimization.sideEffects` of. the Webpack config but I guess this is how it is.

Anyway, thanks so much for your help!
